server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -4947,6 +4974,66 @@ public void testKeepMinRetainedSeqNoByMergePolicy() throws IOException {
+                        .scorer(leaf).iterator();
+                    int doc;
+                    while((doc = iterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                        assertThat(seqNoDV.advanceExact(doc), equalTo(true));
+                        if (updatedByDV.advanceExact(doc)) {
+                            assertThat(updatedByDV.longValue(), greaterThanOrEqualTo(0L));
+                            updates.put(seqNoDV.longValue(), updatedByDV.longValue());
+                        }
+                    }
+                }
+            }
+            return updates;
+        };
+        Settings.Builder settings = Settings.builder()
+            .put(defaultSettings.getSettings())
+            .put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true);
+        final IndexMetaData indexMetaData = IndexMetaData.builder(defaultSettings.getIndexMetaData()).settings(settings).build();
+        final IndexSettings indexSettings = IndexSettingsModule.newIndexSettings(indexMetaData);
+        // On replica
+        AtomicLong globalCheckpoint = new AtomicLong(SequenceNumbers.NO_OPS_PERFORMED);
+        try (Store store = createStore();
+             InternalEngine engine = createEngine(config(indexSettings, store, createTempDir(),
+                 newMergePolicy(), null, null, globalCheckpoint::get))) {
+            engine.index(replicaIndexForDoc(createParsedDoc("1", null), 1, 0, false));
+            engine.index(replicaIndexForDoc(createParsedDoc("1", null), 2, 1, false));
+            // IndexWriter allows softUpdateDocuments soft-deleted docs again until it's refreshed.
+            engine.refresh("test");
+            engine.delete(replicaDeleteForDoc("1", 4, 3, threadPool.relativeTimeInMillis()));
+            engine.index(replicaIndexForDoc(createParsedDoc("1", null), 3, 2, false));
+            engine.index(replicaIndexForDoc(createParsedDoc("1", null), 5, 4, false));
+            Map<Long, Long> seqNos = readUpdatedBySeqNos.apply(engine);
+            assertThat(seqNos, hasEntry(0L, 1L)); // 0 -> 1
+            assertThat(seqNos, hasEntry(1L, 4L)); // 1 -> 3 -> 4
+            assertThat(seqNos, hasEntry(2L, 3L)); // 2 -> 3 (stale)
+        }
+        // On primary
+        try (Store store = createStore();
+             InternalEngine engine = createEngine(config(indexSettings, store, createTempDir(),
+                 newMergePolicy(), null, null, globalCheckpoint::get))) {
+            engine.index(indexForDoc(createParsedDoc("1", null)));
+            engine.index(indexForDoc(createParsedDoc("1", null)));
+            engine.delete(new Engine.Delete("test", "1", newUid("1"), primaryTerm.get()));
+            engine.index(indexForDoc(createParsedDoc("1", null)));
+            engine.refresh("test");
+            Map<Long, Long> seqNos = readUpdatedBySeqNos.apply(engine);
+            assertThat(seqNos, hasEntry(0L, 2L)); // 0 -> 1 -> 2
+            assertThat(seqNos, hasEntry(1L, 2L)); // 1 -> 2
```

Comment:
assert that the total size is what we expect?

Reply:
Done